spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/actions/RemoveDanglingDeletesSparkAction.java

```diff
@@ -0,0 +1,227 @@
+ *       the same partition
+ * </ul>
+ */
+public class RemoveDanglingDeletesSparkAction
+    extends BaseSnapshotUpdateSparkAction<RemoveDanglingDeletesSparkAction>
+    implements RemoveDanglingDeleteFiles {
+
+  private final Table table;
+
+  protected RemoveDanglingDeletesSparkAction(SparkSession spark, Table table) {
+    super(spark);
+    this.table = table;
+  }
+
+  @Override
+  protected RemoveDanglingDeletesSparkAction self() {
+    return this;
+  }
+
+  @Override
+  public Result execute() {
+    if (table.specs().size() == 1 && table.spec().isUnpartitioned()) {
+      // ManifestFilterManager already performs this table-wide on each commit
+      return RemoveDanglingDeleteFilesActionResult.empty();
+    }
+
+    String desc = String.format("Remove dangling delete files for %s", table.name());
+    JobGroupInfo info = newJobGroupInfo("REWRITE-MANIFESTS", desc);
+    return withJobGroupInfo(info, this::doExecute);
+  }
+
+  private Result doExecute() {
+    Dataset<Row> entries =
+        loadMetadataTable(table, ENTRIES)
+            .filter("status < 2") // live entries
+            .selectExpr(
+                "data_file.partition as partition",
+                "data_file.spec_id as spec_id",
+                "data_file.file_path as file_path",
+                "data_file.content as content",
+                "data_file.file_size_in_bytes as file_size_in_bytes",
+                "data_file.record_count as record_count",
+                "sequence_number");
+
+    DeleteFiles deleteFiles = table.newDelete();
+    List<DeleteFile> toRemove = withReusableDS(entries, this::danglingDeletes);
+    toRemove.forEach(deleteFiles::deleteFile);
+    deleteFiles.commit();
+
+    return new RemoveDanglingDeleteFilesActionResult(toRemove);
+  }
+
+  /**
+   * Calculate dangling delete files
+   *
+   * <ul>
+   *   <li>Group all files by partition, calculate the minimum data file sequence number in each.
+   *   <li>For each partition, check if any position delete files have a sequence number less than
+   *       the partition's min_data_sequence_number
+   *   <li>For each partition, check if any equality delete files have a sequence number less than
+   *       or equal to the partition's min_data_sequence_number
+   * </ul>
+   *
+   * @param entries dataset of file entries, marked by content (0 for data, 1 for posDeletes, 2 for
+   *     eqDeletes)
+   * @return list of dangling delete files
+   */
+  private List<DeleteFile> danglingDeletes(Dataset<Row> entries) {
+    List<DeleteFile> removedDeleteFiles = Lists.newArrayList();
+
+    // Minimum sequence number of data files in each partition
+    Dataset<Row> minDataSeqNumberPerPartition =
+        entries
+            .filter("content == 0") // data files
+            .groupBy("partition", "spec_id")
+            .agg(min("sequence_number"))
+            .toDF("partition", "spec_id", "min_data_sequence_number");
+
+    // Dangling position delete files
+    Column joinCond =
```

Comment:
I can't reproduce it on my own machine, but on build machine I get the error if I try the cleaner: Dataset.join(ds, Seq<String> columns) 

```error: no suitable method found for join(Dataset<Row>,Buffer<String>)``` see: https://github.com/apache/iceberg/actions/runs/3913210764/jobs/6688831726

So I make an explicit join condition, like RewriteManifestFileSparkAction